lib/fog/compute/google/requests/reset_windows_password.rb

```diff
@@ -0,0 +1,143 @@
+        # @param user [String] the user whose password should be reset
+        #
+        # @return [String] new password
+        #
+        # @see https://cloud.google.com/compute/docs/instances/windows/automate-pw-generation
+        def reset_windows_password(instance, zone, user)
+         
+          # Pull the e-mail address of user authenticated to API 
+          email = @compute.request_options.authorization.issuer
+          
+          # Create a new key
+          key = OpenSSL::PKey::RSA.new(2048)
+          modulus, exponent = get_modulus_exponent_in_base64(key)
+
+					# Get Old Metadata
+          old_metadata = get_server(instance,zone).metadata
+          fingerprint = old_metadata.fingerprint
+
+          # Create JSON Object with needed information
+          metadata_entry = get_json_string(user, modulus, exponent, email)
+
+          # Create new metadata object
+          new_metadata = update_windows_keys(old_metadata, metadata_entry)
+
+          # Set metadata on instance
+          result = update_instance_metadata(instance, zone, fingerprint, new_metadata)
+
+          # Need to wait for new metadata to take effect.
+          # TODO: Find a way to poll serial output to wait for
+          # it there.
+          sleep(60)
+
+          # Get encrypted password from Serial Port 4 Output
+          serial_port_output = get_serial_port_four_output(instance,zone)
+          enc_password = get_encrypted_password_from_serial_port(serial_port_output, modulus)
+          password = decrypt_password(enc_password, key)
+
+          return password
+        end
+
+        def get_modulus_exponent_in_base64(key)
+					mod = [key.n.to_s(16)].pack('H*').strip
+					exp = [key.e.to_s(16)].pack('H*').strip
+					modulus = Base64.strict_encode64(mod).strip
+					exponent = Base64.strict_encode64(exp).strip
+					return modulus, exponent
+        end
+
+        def get_expiration_time_string()
+          utc_now = Time.now.utc
+          expire_time = utc_now + 5 * 60
+          return expire_time.strftime("%Y-%m-%dT%H:%M:%SZ")
+        end
+
+        def get_json_string(user, modulus, exponent, email)
+          expire = get_expiration_time_string()
+          data = { 'userName': user,
+                  'modulus': modulus,
+                  'exponent': exponent,
+                  'email': email,
+                  'expireOn': expire }
+          return ::JSON.dump(data)
+        end
+
+        def update_windows_keys(old_metadata, metadata_entry)
+          # TODO: Deep copy old metadata
+          new_metadata = [{
+            'key': "windows-keys",
+            'value': metadata_entry
+          }]
+          return new_metadata
+        end
+
+        def update_instance_metadata(instance, zone, fingerprint, new_metadata)
+          return set_server_metadata(instance, zone, fingerprint, new_metadata)
+        end
+
+        def get_serial_port_four_output(instance, zone)
```

Comment:
Why are you using a separate method here?

Asking since ruby doesn't need an explicit return, so the only thing we're doing here really is instantiating the `port` variable.

Reply:
This was a hold over from the Google example code.  I've refactored this into a direct call.